File: src/Hemordna.Client/Support/Speech.cs
using Microsoft.JSInterop;

namespace Hemordna.Client.Support;

/// <summary>
/// "Läs upp" (fokusläget, MinDag.razor) - a thin wrapper over the Web Speech API
/// (<c>js/speech.js</c>). No per-device or per-member state to persist (unlike
/// <see cref="Theme"/>/<see cref="CalmScreen"/>/<see cref="StartedTask"/>): every press starts a
/// fresh utterance, and there is nothing to remember between visits.
/// </summary>
internal static class Speech
{
    /// <summary>
    /// "Kan den här enheten tala?" - never throws because the module could not be loaded (an old
    /// browser, an interrupted download). MinDag.razor's OnInitializedAsync calls this
    /// unprotected, and a failed import must not block the whole page the way it once did - see
    /// HouseholdRealtimeClient.ConnectAsync's own remarks for the same class of bug. <c>false</c>
    /// here means "could not confirm it works", not "this device is proven unable to speak" - the
    /// natural answer to the question either way, but worth reading correctly: a later visit could
    /// still succeed. An <see cref="OperationCanceledException"/> from a shutdown in progress is
    /// let through unchanged, since that is not a module-loading failure.
    /// </summary>
    public static async Task<bool> IsAvailableAsync(IJSRuntime js)
    {
        try
        {
            await using var module = await js.InvokeAsync<IJSObjectReference>("import", "./js/speech.js");
            return await module.InvokeAsync<bool>("isAvailable");
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception)
        {
            return false;
        }
    }

    /// <summary>Same non-throwing contract as <see cref="IsAvailableAsync"/>, for the same reason
    /// - a failed module import answers "could not confirm a Swedish voice exists" as
    /// <c>false</c>, not as a thrown exception.</summary>
    public static async Task<bool> HasSwedishVoiceAsync(IJSRuntime js)
    {
        try
        {
            await using var module = await js.InvokeAsync<IJSObjectReference>("import", "./js/speech.js");
            return await module.InvokeAsync<bool>("hasSwedishVoice");
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (Exception)
        {
            return false;
        }
    }

    /// <summary>Cancels whatever was already playing, same as the button's own "one at a time"
    /// behaviour. <paramref name="selfRef"/> receives the JS side's "OnSpeechEnded" callback once
    /// the utterance actually finishes (or errors) - a caller-owned reference, not created here,
    /// since it needs to outlive this one call.</summary>
    public static async Task SpeakAsync<T>(IJSRuntime js, DotNetObjectReference<T> selfRef, string text)
        where T : class
    {
        await using var module = await js.InvokeAsync<IJSObjectReference>("import", "./js/speech.js");
        await module.InvokeVoidAsync("speak", selfRef, text);
    }

    public static async Task StopAsync(IJSRuntime js)
    {
        await using var module = await js.InvokeAsync<IJSObjectReference>("import", "./js/speech.js");
        await module.InvokeVoidAsync("stop");
    }
}

File: src/Hemordna.Client/wwwroot/js/speech.js
// "Läs upp" (MinDag.razor, fokusläget) - a thin wrapper over the Web Speech API. No state kept
// here beyond the one in-flight utterance; a new speak() always cancels whatever was playing.

export function isAvailable() {
    return 'speechSynthesis' in window;
}

// Voices can load asynchronously in some browsers (an empty list right after startup does not
// mean "no voices ever") - an empty list here is read as "cannot tell yet", not "no Swedish
// voice", so a device that just has not finished loading its voice list does not get shown a
// caveat that may turn out to be wrong a moment later.
export function hasSwedishVoice() {
    if (!isAvailable()) {
        return false;
    }

    const voices = window.speechSynthesis.getVoices();
    if (voices.length === 0) {
        return true;
    }

    return voices.some(voice => voice.lang && voice.lang.toLowerCase().startsWith('sv'));
}

export function speak(dotNetRef, text) {
    window.speechSynthesis.cancel();

    const utterance = new SpeechSynthesisUtterance(text);
    utterance.lang = 'sv-SE';
    utterance.rate = 0.95;
    utterance.onend = () => dotNetRef.invokeMethodAsync('OnSpeechEnded');
    utterance.onerror = () => dotNetRef.invokeMethodAsync('OnSpeechEnded');

    window.speechSynthesis.speak(utterance);
}

export function stop() {
    window.speechSynthesis.cancel();
}
